File: PolarsVsSpark/spark/bench1.scala
import org.apache.spark.sql.functions._
import org.apache.spark.sql.{DataFrame, SparkSession}

object bench1 {
  def read_csvs(spark: SparkSession, path: String): DataFrame =
    spark.read
      .option("recursiveFileLookup", "true")
      .option("header", "true")
      .csv(path)

  def write_parquets(df: DataFrame, path: String): Unit =
    df.write
      .partitionBy("date")
      .format("parquet")
      .mode("overwrite")
      .save(path)

  def calculate_metrics(df: DataFrame): DataFrame =
    df
      .groupBy(
        col("date"),
        year(col("date")).alias("year"),
        month(col("date")).alias("month"),
        dayofmonth(col("date")).alias("day"),
        col("model")
      )
      .agg(sum("failure").alias("failures"))

  def main(args: Array[String]): Unit = {
    val t1 = System.nanoTime()

    val read_path = "root/input/*.csv"
    val metrics_write_path = "root/parquets/parquets_metrics"
    val spark = SparkSession.builder.getOrCreate()

    val raw_data = read_csvs(spark, read_path)
    val selected_data = raw_data.select("date", "model", "failure")
    val metrics = calculate_metrics(selected_data)
    write_parquets(metrics, metrics_write_path)

    val t2 = System.nanoTime()
    println("Time to run Apache Spark pipeline : " + (t2 - t1) / 1e9 + "s")
    spark.stop()
  }
}

File: PolarsVsSpark/polars/bench1.py
import polars as pl
from datetime import datetime


def read_csvs(path: str):
    lazy_df = pl.scan_csv(path)
    return lazy_df


def calculate_metrics(df):
    metrics = \
        df.group_by(
            pl.col('date').cast(pl.Date), 
            pl.col('date').cast(pl.Date).dt.year().alias('year'),
            pl.col('date').cast(pl.Date).dt.month().alias('month'),
            pl.col('date').cast(pl.Date).dt.day().alias('day'),
            pl.col('model')
            ).agg(pl.sum("failure").alias("failures")
        )
    return metrics


def write_parquets(df, path: str) -> None:
    df.sink_parquet(path, compression='snappy')


def main():
    t1 = datetime.now()

    read_path = 'root/input/*.csv'
    metrics_write_path = 'root/parquets/parquets_metrics2'

    df = read_csvs(read_path)
    df = df.select('date', 'model', 'failure')
    metrics = calculate_metrics(df)
    write_parquets(metrics, metrics_write_path)

    t2 = datetime.now()
    print(f"Time to run polars pipeline : {t2 - t1}")


if __name__ == "__main__":
    main()